### ISSUE TITLE
The order of destruction is wrong

### ISSUE THREAD

**Issue by gracicot** [gracicot/kangaru#69]
When using service that refers on each other on destruction, you're up for a bad time un kangaru:
```c++
struct A { int boom; };
struct B {
    A& a;

    ~B() {
        if (a.boom) {} // boom
    }
};

struct AService : kgr::single_service<A> {};
struct BService : kgr::single_service<B, kgr::dependency<AService>> {};

int main() {
    kgr::container{}.service<BService>(); // boom
}
```

This code uses freed memory because it deletes `A` before `B`, which we think it's wrong, the container should guarantee that since `B` depends on `A`, `A` should be useable throughout the whole lifetime of `B`.

**Comment by gracicot** [comment]
Changing behaviour like that is certainly breaking. Code that worked could start crashing if it relied on the old destruction order. A switch should be provided to activate the new behaviour.

**Comment by gracicot** [comment]
This is fix in `dev-4.2.x`

**Comment by gracicot** [comment]
The fix has been backported to `v4.0.4` and `v4.1.3`
